halo2-base/src/gates/range.rs

```diff
@@ -472,6 +494,33 @@ impl<F: ScalarField> RangeChip<F> {
     pub fn default(lookup_bits: usize) -> Self {
         Self::new(RangeStrategy::Vertical, lookup_bits)
     }
+
+    /// Bitwise right rotate a by <bit> bits. This function should never be called directly
+    /// because const bitwise rotation must be determined at compile time.
+    ///
+    /// Assumes 'a' is a <num_bits> bit integer and 0 < <num_bits> <= 128.
+    fn const_right_rotate_internal(
+        &self,
+        ctx: &mut Context<F>,
+        a: AssignedValue<F>,
+        bit: usize,
+        num_bits: usize,
+    ) -> AssignedValue<F> {
+        assert!(0 < num_bits && num_bits <= 128);
+        assert!(a.value().cmp(&F::from_u128(u128::MAX >> (128 - num_bits))).is_le());
+        // Add a constrain a = l_witness << bit | r_wintess
+        let val = a.value().get_lower_128();
+        let val_l = val >> bit;
+        let val_r = val - (val_l << bit);
+        let l_witness = ctx.load_witness(F::from_u128(val_l));
```

Comment:
A bit low level, but an optimization is to still use `Witness(F::from_u128(val_l))` and then fetch `l_witness = ctx.get(-3), r_witness = ctx.get(-4)`